### ISSUE TITLE
Zero grad in residual vq

### ISSUE THREAD

**Issue by npuichigo** [facebookresearch/encodec#25]
## 🐛 Bug Report
Zero grad in second residual vq as mentioned here (https://github.com/lucidrains/vector-quantize-pytorch/issues/33)
https://github.com/facebookresearch/encodec/blob/194329839fd812433992272fc5e7a889176e6fd1/encodec/quantization/core_vq.py#L336

The fix link is https://github.com/lucidrains/vector-quantize-pytorch/commit/ecf2f7c7c7701029488500b6b3d46b7c895063bf

**Comment by npuichigo** [comment]
@adefossez

**Comment by adefossez** [comment]
Thanks for bringing that out!

It seem like this won't impact the Straight-Through-Estimator gradient for the Encoder, but will kill the commitment loss for all residual VQ but the first one right ?

**Comment by npuichigo** [comment]
It seems so. But I'm not sure how much it affects the final result.

**Comment by adefossez** [comment]
I'm a bit reluctant on introducing a change we haven't tested in this codebase, as it could change the best hyper params etc. I can add a warning however if the model is used in training mode pointing to this issue.

**Comment by adefossez** [comment]
why did you put `0 *  q1.sum()` ? that is what is breaking the STE gradient. With the current code `d q1 / d x = Id` and `d q_i d / x = 0` for all `i > 1`, which is okay as the overall gradient `d (sum q_i) / d x = Id` which is what we want. The only thing that is impacted in the commitment loss.

**Comment by DingWeiPeng** [comment]
@adefossez @cantabile-kwok 

If  residual = residual - quantized ,  then the second codebook can update with gradient but it can not afffect the first codebook.
If residual = residual - quantized.detach(), then the second codebook's gradient will affect the fisrt codebook.

In core_vq.py, there is the following code in VectorQuantization Class : 
![image](https://github.com/facebookresearch/encodec/assets/46740555/5094c4c4-dd2c-4a8c-8e6d-ca74041f442d)

Now there is the following code in the ResidualVectorQuantization Class
![image](https://github.com/facebookresearch/encodec/assets/46740555/05f758b7-7972-420c-bf72-c48c18a78a94)

So, this problem equals to the following problem. The following code snippet may illustrate this:

'''
import torch
def quantize(x, codebook):
    diff = codebook - x  # (n_code, dim)
    mse = (diff**2).sum(1)
    idx = torch.argmin(mse)
    return codebook[idx]

dim = 5
x = torch.randn(1, dim, requires_grad=True)
codebook1 = torch.randn(10, dim)
codebook2 = torch.randn(10, dim)

q1 = quantize(x, codebook1)  # quantize x with first codebook
q1 = x + (q1 - x).detach()  # transplant q1's gradient to x
residual = x - q1.detach()  # detach q1 or not may make a difference. Compute residual for next level quantizing
q2 = quantize(residual, codebook2)  # quantize residual with second codebook
q2 = residual + (q2 - residual).detach()  # transplant q2's gradient to residual

loss = 1*q2.sum()  # loss is a function of q1 and q2, now it is independent of q1.
loss.backward()
print(x.grad)
'''

if residual = x-q1, x.grad = 0,
if residul = x-q1.detach(), x.grad = tensor([[1., 1., 1., 1., 1.]])

**Comment by DingWeiPeng** [comment]
@adefossez May I ask if you used the EMA method to update the codebook in the rvq algorithm?
